pinot-orc/src/main/java/org/apache/pinot/orc/data/readers/ORCRecordReader.java

```diff
@@ -0,0 +1,207 @@
+
+  private void init(String inputPath, Schema schema) {
+    Configuration conf = new Configuration();
+    LOGGER.info("Creating segment for {}", inputPath);
+    try {
+      Path orcReaderPath = new Path(LOCAL_FS_PREFIX + inputPath);
+      LOGGER.info("orc reader path is {}", orcReaderPath);
+      _reader = OrcFile.createReader(orcReaderPath, OrcFile.readerOptions(conf));
+      _orcSchema = _reader.getSchema();
+      LOGGER.info("ORC schema is {}", _orcSchema.toJson());
+
+      _pinotSchema = schema;
+      if (_pinotSchema == null) {
+        LOGGER.warn("Pinot schema is not set in segment generator config");
+      }
+      _recordReader = _reader.rows(_reader.options().schema(_orcSchema));
+    } catch (Exception e) {
+      LOGGER.error("Caught exception initializing record reader at path {}", inputPath);
+      throw new RuntimeException(e);
+    }
+
+    // Create a row batch with max size 1
+    _reusableVectorizedRowBatch = _orcSchema.createRowBatch(1);
+  }
+
+  @Override
+  public void init(SegmentGeneratorConfig segmentGeneratorConfig) {
+    init(segmentGeneratorConfig.getInputFilePath(), segmentGeneratorConfig.getSchema());
+  }
+
+  @Override
+  public boolean hasNext() {
+    try {
+      return _recordReader.getProgress() != 1;
+    } catch (IOException e) {
+      LOGGER.error("Could not get next record");
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public GenericRow next()
+      throws IOException {
+    return next(new GenericRow());
+  }
+
+  @Override
+  public GenericRow next(GenericRow reuse)
+      throws IOException {
+    _recordReader.nextBatch(_reusableVectorizedRowBatch);
+    fillGenericRow(reuse, _reusableVectorizedRowBatch);
+    return reuse;
+  }
+
+  private void fillGenericRow(GenericRow genericRow, VectorizedRowBatch rowBatch) {
+    // ORC's TypeDescription is the equivalent of a schema. The way we will support ORC in Pinot
+    // will be to get the top level struct that contains all our fields and look through its
+    // children to determine the fields in our schemas.
+    if (_orcSchema.getCategory().equals(TypeDescription.Category.STRUCT)) {
+      for (int i = 0; i < _orcSchema.getChildren().size(); i++) {
+        // Get current column in schema
+        TypeDescription currColumn = _orcSchema.getChildren().get(i);
+        String currColumnName = _orcSchema.getFieldNames().get(i);
+        if (!_pinotSchema.getColumnNames().contains(currColumnName)) {
+          LOGGER.warn("Skipping column {} because it is not in pinot schema", currColumnName);
+          continue;
+        }
+        int currColRowIndex = currColumn.getId();
+        // Struct is top level, so the id of the struct is 0. However, the children start from 1+, etc, so we need to
+        // subtract one since the row batch we get has only children column vectors
+        ColumnVector vector = rowBatch.cols[currColRowIndex - 1];
```

Comment:
You don't need to use `getId()` for indexing into the returned batch. The ids are an internal detail of ORC and reference individual streams in the serialized file. The VectorizedRowBatch that's returned to you will have exactly the fields that you specified when you set the schema in your ReaderOptions (or the file schema if you don't set it) and they will be in the same order as that schema. Therefore, you can just use `rowBatch.cols[i]` to get what you want.

Reply:
That `currColRowIndex - 1` happens to work is only because you've been using flat schemas.

Reply:
Thank you so much! Changed